File: src/infinicore/ops/hemm/hemm.cc
#include "infinicore/ops/hemm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Hemm);

Hemm::Hemm(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, b, alpha, beta, c);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), a, b, alpha, beta, c, side, uplo);
}

void Hemm::execute(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Hemm, a, b, alpha, beta, c, side, uplo);
}

void hemm_(const Tensor &a, const Tensor &b, const Tensor &alpha, const Tensor &beta, Tensor c, int side, int uplo) {
    Hemm::execute(a, b, alpha, beta, c, side, uplo);
}

} // namespace infinicore::op

File: python/infinicore/ops/hemm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def hemm(
    a: Tensor,
    b: Tensor,
    alpha: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    side: int = 0,
    uplo: int = 0,
):
    _infinicore.hemm_(
        a._underlying,
        b._underlying,
        alpha._underlying,
        beta._underlying,
        out._underlying,
        side,
        uplo,
    )

    return out
